### ISSUE TITLE
Issue with single attribute variation display (2.4.7)

### ISSUE THREAD

**Issue by janiszalitis** [woothemes/woocommerce#9287]
Hi!

After upgrading to 2.4.7 I have an issue with single attribute variations.

I have products with a single attribute that defines a variation - size ("Izmērs").

When I open product page, the dropdown is cleared of values by JavaScript, despite having the values on output, only option "Choose an option" remains in the select box. 
The "clear selection" button is also hidden.
Here's how I detected the problem:
1. Tried opening the same product page with JavaScript OFF. The dropdown showed variations just fine.
2. Switched JavaScript ON and refreshed. Variations disappeared.
3. Renamed the folder /wp-content/plugins/woocommerce/assets/js/frontend/ so that the site would not load any script from it. Variations became visible again in the dropdown.
4. Renamed the folder back to normal. Refreshed. Variations disappeared.
5. Renamed the file: /wp-content/plugins/woocommerce/assets/js/frontend/add-to-cart-variation.min.js so that the site would not find it. Variations visible.
6. Renamed it back. Refreshed. Variations disappeared.
7. Swapped the "add-to-cart-variation.min.js" with "add-to-cart-variation.js", to be able to edit code. Assumed both files identical. Variations not visible on refresh (expected behaviour).
8. Commented out parts of code until finally came down to adding "return false;" at line #309. Variations now visible. 

I then noticed the call @ line #332 and realized this is the likely culprit:
<pre>var matching_variations = wc_variation_form_matcher.find_matching_variations( $product_variations, current_settings );</pre>

I added this code right after that call:
<pre>console.info('matching_variations: ' + matching_variations)</pre>

Which gave me this:
![screen shot 2015-10-05 at 20 52 58](https://cloud.githubusercontent.com/assets/1185430/10289256/1e3398f4-6ba6-11e5-923a-2373523736ba.jpg)

I immediately understood that something is either wrong with the matching function or it's inputs.

I proceeded to the matching function below, found @ line #467:
<pre>
find_matching_variations: function( product_variations, settings ) {
			var matching = [];			
			for ( var i = 0; i < product_variations.length; i++ ) {
				var variation    = product_variations[i];

				if ( wc_variation_form_matcher.variations_match( variation.attributes, settings ) ) {
					matching.push( variation );
				}
			}
			return matching;
		},
</pre>

I then modified it as follows:
<pre>
var wc_variation_form_matcher = {
		find_matching_variations: function( product_variations, settings ) {
			var matching = [];
			
			console.info('product_variations: ' + product_variations)
			
			for ( var i = 0; i < product_variations.length; i++ ) {
				var variation    = product_variations[i];

				if ( wc_variation_form_matcher.variations_match( variation.attributes, settings ) ) {
					matching.push( variation );
				}
			}
			return matching;
		},
</pre>

Which gave me this:
![screen shot 2015-10-05 at 21 11 39](https://cloud.githubusercontent.com/assets/1185430/10289311/63ecb47a-6ba6-11e5-90ad-b1ac714ea74e.jpg)

Finally, I modified it as follows (note the 2nd console.info call):
<pre>
var wc_variation_form_matcher = {
		find_matching_variations: function( product_variations, settings ) {
			var matching = [];
			
			console.info('product_variations: ' + product_variations)
			
			for ( var i = 0; i < product_variations.length; i++ ) {
				var variation    = product_variations[i];
				
				console.info('variation: ' + variation)

				if ( wc_variation_form_matcher.variations_match( variation.attributes, settings ) ) {
					matching.push( variation );
				}
			}
			return matching;
		},
</pre>

Which gave me this:

![screen shot 2015-10-05 at 21 12 46](https://cloud.githubusercontent.com/assets/1185430/10289415/d2b5fb3c-6ba6-11e5-9022-e0ce88029d80.jpg)

It appears that the matching function is trying to use a string of malformed JSON as an object and runs into problems.

Further fiddling led me to this, at which point I realized that the problem may be with the double quotes:
![screen shot 2015-10-05 at 21 27 47](https://cloud.githubusercontent.com/assets/1185430/10289635/0f93984c-6ba8-11e5-8aa5-cadcc7bdffff.jpg)

This is how the contents of data-product_variations parameter for corresponding form look like (note the sku value):
<pre>[{"variation_id":16696,"variation_is_visible":true,"variation_is_active":true,"is_purchasable":true,"display_price":359.2,"display_regular_price":449,"attributes":{"attribute_pa_izmers":"17"},"image_src":"","image_link":"","image_title":"","image_alt":"","price_html":"","availability_html":"<p class=\"stock in-stock\">Vairumtirdzniec\u012bbas noliktav\u0101: 2<\/p>","sku":"C Access WLS 26 wht 15-17"","weight":" kg","dimensions":"","min_qty":1,"max_qty":2,"backorders_allowed":false,"is_in_stock":true,"is_downloadable":false,"is_virtual":false,"is_sold_individually":"no","variation_description":""}]</pre>

And this is how the code that is supposed to escape these values looks like:
<pre>data-product_variations="< ? php echo esc_attr( json_encode( $available_variations ) ) ? ></pre>
From: /wp-content/plugins/woocommerce/templates/single-product/variable.php, line #19

Any suggestions how to fix this?
I could change the name of the product (or rather several hundred), but I feel it shouldn't be required. 
There must be thousands of products in hundreds of stores, that have inches in their title field.

Thank you for great plugin!

**Comment by mikejolley** [comment]
Core escapes these correctly https://dl.dropboxusercontent.com/s/5qztq567oblgmzo/2015-10-06%20at%2013.29.png?dl=0

You likely have a plugin breaking this; I've seen one of the translation plugins do this (changes the way json_encode works).

**Comment by janiszalitis** [comment]
The root of the problem, turns out, is that in the database the SKU looks like this:
<pre>C Aim 26 lime/blue 15-18&quot;</pre>
I am not sure how it got to be like, but let's assume such an input is possible (even if someone just typed the entity code manually, which I didn't do).

What next happens is, json_encode gets to encode this, which it does correctly.
The result looks like this:
<pre>"sku":"C Aim 26 lime/blue 15-18&quot;"</pre>

Then this is passed to esc_attr, and the result looks like this:
<pre>&quot;sku&quot;:&quot;C Aim 26 lime/blue 15-18&quot;&quot;</pre>

This is what I added to my functions.php and it immediately fixed the problem:
<pre>add_filter('woocommerce_available_variation', 'fans_woocommerce_available_variation', 99, 3);
function fans_woocommerce_available_variation($data, $obj, $variation)
{
	if(!empty($data['sku'])) $data['sku'] = str_replace('&quot;', '"', $data['sku']);
	
	return $data;
}</pre>

I see 3 possible solutions:
1. Validate value upon data insertion (it's possible this is happening already, I did not check, but it would not be 100% safe anyway, just in case data still get skewed somehow, like by 3rd party plugin or a future bug in Woocommerce) and convert the entities to their "normal" representation.

2. Validate value upon data retrieval, e.g. in the get_sku function. There's also a filter for that - "woocommerce_get_sku". This is what I did above. It should be 100% safe, at least within confines of Woocommerce and any 3rd parties who use Woocommerce methods for data retrieval.

3. Fix/hook/override(or don't use) esc_attr so that it is "aware" of situations where in a string there are unopened or unended quotes.


For the record I also disabled all plugins except Woocommerce on a test installation. 
Obviously, it didn't fix anything, although I suspect one of those plugins (WP All Import) could have inserted the quot; value. 
I didn't check the code for insertion WP All Import or data validation in Woocommerce and it's possibly irrelevant as this is probably not the only case where using esc_attr can lead to such a problem.


Hopefully you either adopt some kind of solution or this helps someone to not spend several hours scratching head, like I did. :)
